crates/composition/src/composition_ir.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    push_and_return_id,
+    subgraphs::{self, StringWalker},
+};
+use grafbase_federated_graph as federated;
+use std::collections::{BTreeSet, HashMap};
+
+/// The intermediate representation of the schema that is produced by composition. This data
+/// structure is distinct from [FederatedGraph](grafbase_federated_graph::FederatedGraph) because
+/// it is accumulated out of order during composition. Only after all composed top-level
+/// definitions (objects, interfaces, scalars, enums, unions) are known can we construct
+/// [federated::FieldTypeId]s and [federated::ObjectId]s for union members for the federated graph.
+///
+/// This is a **write only** data structure during composition. The source of truth for the
+/// contents of the federated graph is the subgraphs.
+#[derive(Default)]
+pub(crate) struct CompositionIr {
+    pub(crate) field_type_kinds_map: HashMap<subgraphs::StringId, federated::FieldTypeKind>,
+
+    pub(crate) objects: Vec<federated::Object>,
+    pub(crate) interfaces: Vec<federated::Interface>,
+    pub(crate) unions: Vec<federated::Union>,
+    pub(crate) enums: Vec<federated::Enum>,
+    pub(crate) scalars: Vec<federated::Scalar>,
+    pub(crate) input_objects: Vec<federated::InputObject>,
+
+    pub(crate) strings: StringsIr,
+    pub(crate) fields: Vec<FieldIr>,
+    pub(crate) union_members: BTreeSet<(subgraphs::StringId, subgraphs::StringId)>,
+}
+
+impl CompositionIr {
+    pub(crate) fn insert_enum(&mut self, enum_name: StringWalker<'_>) -> federated::EnumId {
+        let name = self.insert_string(enum_name);
+        push_definition_and_return_id(
```

Comment:
Not too convinced by this util function, maybe we can make these calls more readable.

Reply:
Yeah.  Maybe just because I'm unfamiliar with this code, but it's not really clear to me what this code is doing.

`push_definition_and_return_id` takes 4 arguments but is only 3 lines long - feels like it might be better to just write out the code (though I could be wrong)

Reply:
Went on and did that, I think it does read better.